test/unit/Monitoring/NullMetricsAgentTest.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * @author    Craig Gosman <craig@ingenerator.com>
+ * @licence   proprietary
+ */
+
+
+namespace test\unit\Ingenerator\PHPUtils\unit\Monitoring;
+
+
+use Ingenerator\PHPUtils\Monitoring\MetricId;
+use Ingenerator\PHPUtils\Monitoring\NullMetricsAgent;
+use PHPUnit\Framework\TestCase;
+
+class NullMetricsAgentTest extends TestCase
+{
+    public function test_it_is_initialisable()
+    {
+        $this->assertInstanceOf(NullMetricsAgent::class, $this->newSubject());
+    }
+
+    /**
+     * @testWith [false]
+     *           [true]
+     *           ["I'm a string"]
+     *
+     * @param mixed $callback
+     */
+    public function test_timeOperation_calls_callback($callback)
+    {
+        $subject = $this->newSubject();
+        $this->assertSame($callback, $subject->timeOperation(function () use ($callback) { return $callback; }));
+    }
+
+    public function test_timeOperation_creates_metric()
+    {
+        $subject = $this->newSubject();
+        $metric  = $subject->timeOperation(function ($metric) { return $metric; }, 'queries', 'mysql');
+        /** @var $metric MetricId */
+        $this->assertInstanceOf(MetricId::class, $metric);
+        $this->assertSame('queries', $metric->getMetricName());
+        $this->assertSame('mysql', $metric->getSource());
```

Comment:
I guess to be pedantic this doesn't guarantee that the default is populated before the call and overridable during - hard to see it going wrong really given the simplicity of the call and obv not actually crucial for this agent anyway. But could do:

```php
    public function test_timeOperation_creates_metric()
    {
        $subject = $this->newSubject();
        $metric  = $subject->timeOperation(
            function (MetricId $metric) {
                $this->assertMetricMatches(['name' => 'queries', 'source' => 'mysql'], $metric);
                $metric->setMetricName('queries.failed');
                $metric->setSource('mysql.slave');
            },
            'queries',
            'mysql'
        );
        $this->assertMetricMatches(['name' => 'queries.failed', 'source' => 'mysql.slave'], $metric);
    }

    private function assertMetricMatches(array $expect, MetricId $metric)
    {
        $this->assertSame($expect, ['name' => $metric->getName(), 'source' => $metric->getSource()]);
    }
```

That covers off the type-safety without needing the `@var` tag, gives a more complete diff on failure (is it just one property assignment or did the whole MetricId instance get swapped / go wrong) and it also helps document why we're passing the MetricId into the callback.